### PR TITLE
[Android] fixes for share stats crash on some lollipop devices

### PR DESCRIPTION
**Pull request by alexsafe**
Resolves https://github.com/brave/brave-browser/issues/15507

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Included in the issue description

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java [hunk 1]
```diff
@@ -163,25 +171,84 @@ public static void updateBraveShareStatsLayoutAndShare(View view) {
...
-        sendIntent.putExtra(Intent.EXTRA_TEXT,
-                context.getResources().getString(R.string.brave_stats_share_text));
-        sendIntent.putExtra(Intent.EXTRA_STREAM, uri);
-        sendIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
-        sendIntent.setType("image/text");
-
-        Intent shareIntent = Intent.createChooser(sendIntent, " ");
-        shareIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
-        context.startActivity(shareIntent);
+        Log.d("CRASHING", "shareStatsAction :");
```
Comment (deeppandya) [inline android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java:174]: Do we need this log ?

File: android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java [hunk 2]
```diff
@@ -163,25 +171,84 @@ public static void updateBraveShareStatsLayoutAndShare(View view) {
...
+        try {
+            Context context = ContextUtils.getApplicationContext();
+            Bitmap bmp = convertToBitmap(view);
+            // ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+            // bmp.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream);
+            Log.d("CRASHING", "getting path :");
```
Comment (deeppandya) [inline android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java:180]: can we remove commented code here ?

File: android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java [hunk 3]
```diff
@@ -163,25 +171,84 @@ public static void updateBraveShareStatsLayoutAndShare(View view) {
...
+            String path = "";
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+                path = MediaStore.Images.Media.insertImage(
+                    context.getContentResolver(), bmp, "tempimage", null);Log.d("CRASHING", "path :"+path);
+            } else {
+                storeImage(bmp);
+                path = getOutputMediaFile().getAbsolutePath();
+            }
+            Log.d("CRASHING", "path:"+getOutputMediaFile().getPath());
```
Comment (deeppandya) [inline android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java:191]: Same about logs

File: android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java [hunk 4]
```diff
@@ -163,25 +171,84 @@ public static void updateBraveShareStatsLayoutAndShare(View view) {
...
+        } catch (IOException e) {
+            Log.d("CRASHING", "Error accessing file: " + e.getMessage());
+        }
+    }
+
+        /** Create a File for saving an image or video */
+    private static File getOutputMediaFile(){
+        // To be safe, you should check that the SDCard is mounted
+        // using Environment.getExternalStorageState() before doing this.
+        File mediaStorageDir = new File(Environment.getExternalStorageDirectory()
```
Comment (deeppandya) [inline android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java:233]: Are we checking for storage permission before this code block ? just making sure

**Review by deeppandya — CHANGES_REQUESTED** [review]

File: android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java (hunk 4 above)
Comment (alexsafe) [reply android/java/org/chromium/chrome/browser/brave_stats/BraveStatsUtil.java:233]: yes, before the share call 

https://github.com/brave/brave-core/blob/e963a3c3ad4b3803d0909f1aa44877a9d93adfe9/android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java#L430

File: android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java [hunk 5]
```diff
@@ -426,8 +426,14 @@ public void onClick(View v) {
             @Override
             public void onClick(View v) {
                 mMainLayout.setVisibility(View.GONE);
+                Log.d("CRASHING", "shareImage click ... ");
```
Comment (deeppandya) [inline android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java:429]: do we still need this log ?

File: android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java [hunk 6]
```diff
@@ -426,8 +426,14 @@ public void onClick(View v) {
...
                 if (BraveStatsUtil.hasWritePermission(BraveActivity.getBraveActivity())) {
+                                    Log.d("CRASHING", "shareImage sharing  ... ");
+
                     BraveStatsUtil.shareStats(R.layout.brave_stats_share_layout);
+                } else {
+                                                        Log.d("CRASHING", "shareImage not sharing  ... ");
```
Comment (deeppandya) [inline android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java:435]: Something went wrong here for formatting.

**Review by deeppandya — APPROVED** [review]
LGTM

**Comment by GeetaSarvadnya** [comment]
Verification passed on Gionee s5 version 5.1 running nightly `Bravearm.apk_1.28.7`
- Confirmed there is no crash when clicking on the share icon in the shiled
- Confirmed click on share icon opens up share tooltip to share the stats via various options like (slack, whats up, mail etc...)
- Confirmed stats can be shared via slack and shared stats are shown correctly in slack
- Confirmed stats can be shared via email and shared stats are shown correctly in Gmail
- Confirmed stats can be shared via what's up app and shared stats are shown correctly in whats up app

**Pull request merged**